gtsfm/utils/cycle_consistency.py

```diff
@@ -0,0 +1,233 @@
+    connected to `a` and the nodes connected to `b`.
+
+    Args:
+        i2Ri1_dict: mapping from image pair indices to relative rotation.
+
+    Returns:
+        triplets: 3-tuples of nodes that form a cycle. Nodes of each triplet are provided in sorted order.
+    """
+    # only want to keep the unique ones
+    triplets = set()
+
+    # form adjacency list
+    adj_list = defaultdict(set)
+
+    for (i1, i2), i2Ri1 in i2Ri1_dict.items():
+        if i2Ri1 is None:
+            continue
+
+        adj_list[i1].add(i2)
+        adj_list[i2].add(i1)
+
+    # find intersections
+    for (i1, i2), i2Ri1 in i2Ri1_dict.items():
+        if i2Ri1 is None:
+            continue
+
+        nodes_from_i1 = adj_list[i1]
+        nodes_from_i2 = adj_list[i2]
+
+        node_intersection = (nodes_from_i1).intersection(nodes_from_i2)
+
+        for node in node_intersection:
+            cycle_nodes = tuple(sorted([i1, i2, node]))
+            triplets.add(cycle_nodes)
+
+    return list(triplets)
+
+
+def compute_cycle_error(
+    i2Ri1_dict: Dict[Tuple[int, int], Rot3],
+    cycle_nodes: Tuple[int, int, int],
+    two_view_reports_dict: Dict[Tuple[int, int], TwoViewEstimationReport],
+    verbose: bool = True,
+) -> Tuple[float, Optional[float], Optional[float]]:
+    """Compute the cycle error by the magnitude of the axis-angle rotation after composing 3 rotations.
+
+    Note: i1 < i2 for every valid edge, by construction.
+
+    Args:
+        i2Ri1_dict: mapping from image pair indices to relative rotation.
+        cycle_nodes: 3-tuples of nodes that form a cycle. Nodes of are provided in sorted order.
+        two_view_reports_dict:
+        verbose: whether to dump to logger information about error in each Euler angle
+
+    Returns:
+        cycle_error: deviation from 3x3 identity matrix, in degrees. In other words,
+            it is defined as the magnitude of the axis-angle rotation of the composed transformations.
+        max_rot_error: maximum rotation error w.r.t. GT across triplet edges, in degrees.
+            If ground truth is not known for a scene, None will be returned instead.
+        max_trans_error: maximum translation error w.r.t. GT across triplet edges, in degrees.
+            If ground truth is not known for a scene, None will be returned instead.
+    """
+    cycle_nodes = list(cycle_nodes)
+    cycle_nodes.sort()
+
+    i0, i1, i2 = cycle_nodes
+
+    i1Ri0 = i2Ri1_dict[(i0, i1)]
+    i2Ri1 = i2Ri1_dict[(i1, i2)]
+    i0Ri2 = i2Ri1_dict[(i0, i2)].inverse()
+
+    # should compose to identity, with ideal measurements
+    i0Ri0 = i0Ri2.compose(i2Ri1).compose(i1Ri0)
+
+    I_3x3 = Rot3()
+    cycle_error = comp_utils.compute_relative_rotation_angle(I_3x3, i0Ri0)
+
+    # form 3 edges between fully connected subgraph (nodes i,j,k)
+    e_i = (i0, i1)
+    e_j = (i1, i2)
+    e_k = (i0, i2)
+
+    rot_errors = [two_view_reports_dict[e].R_error_deg for e in [e_i, e_j, e_k]]
```

Comment:
Is it possible for two_view_reports_dict to have an edge in a different order? like (i1, i0)?

Reply:
so it is not, since the keys for `two_view_reports_dict` will be used from the loader class, and in the loader, we always ensure `a < b` for any image indices `a, b`